### ISSUE TITLE
Falling nodes fall infinitely on overhigh collisionbox

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#9293]
##### Minetest version
```
5.1.0
```

##### Summary
When a falling node falls on a node with an overhigh collision box (i.e. upper collision box Y > 1.0), the falling node stops (because it collides), then it teleports a short distance upwards, then falls again, and stops again at the overhigh collisionbox, etc. In other words, the falling node is stuck in an infinite loop. This is really bad.

##### Steps to reproduce
1. Place a node with an overhigh collisionbox
2. Make a node above it fall on it

You can use `devtest` <https://git.minetest.land/Wuzzy/devtest>.

##### Additional context
#6985 attempted to fix this, but failed. The solution in this PR was bad.

**Comment by SmallJoker** [comment]
If it works in ` __builtin:item`, it could be applied the same way for falling nodes.


_inb4 Wuzzy opens another issue about the same bug in dropped items_

**Comment by Wuzzy2** [comment]
The code that triggers the teleportation is in `builtin/game/falling.lua:139`:

```
                 if vector.equals(vel, {x = 0, y = 0, z = 0}) then
                         local npos = self.object:get_pos()
                         self.object:set_pos(vector.round(npos))
                 end
```

If the falling node comes to a standstill, it will teleport 1 unit upwards.
Removing this code is not a fix. The falling node will just stop but not turn into a node.

The code is pretty hacky as well, this check definitely needs to be changed anyway. Checking for 0 speed is not a smart idea. MTG hurls falling nodes upwards with TNT, so a 0 speed can be achieved without collision.

The tricky part here is to figure out where to place the node. Let's discuss potential solutions.
A rejected solution was to make it drop as an item (see linked issue).
A different solution is to scan 2 node lengths below the entity, to better predict when it will hit an overhigh collisionbox. But it might be bad for performance.

Dropped items do not have this teleportation. They stop properly on the collisionbox.

**Comment by HybridDog** [comment]
> The code is pretty hacky as well, this check definitely needs to be changed anyway.

The position rounding is required when the falling node object is blocked by a solid node:
![situation](https://user-images.githubusercontent.com/3192173/72185935-48860180-33f4-11ea-8d27-173dc7b04e32.png)
black dots: integer positions, blue dot: floating point object position, 
red: falling object, green: some solid node

The object always accelerates up or down, so you only need to round the x and z values:
```lua
                 if vector.equals(vel, {x = 0, y = 0, z = 0}) then
                         local npos = self.object:get_pos()
                         npos.x = math.floor(npos.x + 0.5)
                         npos.z = math.floor(npos.z + 0.5)
                         self.object:set_pos(npos)
                 end
```

Btw: Do falling nodes still have the problem where sand can fall into a walkable chest and then this chest and all its contents are removed?
A long time ago, I added [a fix for this (I am not sure if it works)](https://github.com/HybridDog/falling_extras/blob/1d8bff0ba7015261376c5e6387373fb0831b3d5a/init.lua#L138) to the falling_extras mod.

**Comment by Wuzzy2** [comment]
It's not a good idea to use the equality operator on floating-point numbers.

**Comment by hecktest** [comment]
It's acceptable when you're comparing to zero.
